File: Sean的leetcode/Leetcode/Longest Substring with At Most K Distinct Characters.cpp
//Sliding Window
//O(n)
class Solution {
public:
    int lengthOfLongestSubstringKDistinct(string s, int k) {
        int n = s.size();
        if (n == 0 || k == 0) return 0;
        vector<int> cnt(256);
        int distinct = 0;
        int l = 0;
        int r = 0;
        int max_len = 0;
       
        for (; r < n; r++) {
            if (cnt[s[r]] == 0) {
                distinct++;
            }
            cnt[s[r]]++;
            while (distinct > k) {
                cnt[s[l]]--;
                if (cnt[s[l]] == 0) {
                    distinct--;
                }
                l++;
            }
            max_len = max(max_len, r - l + 1);
        }
        return max_len;
    }
};

class Solution {
public:
    int lengthOfLongestSubstringKDistinct(string s, int k) {
        int l = 0;
        int r = 0;
        int maxlen = 0;
        int dist = 0;
        vector<int> cnt(256, 0);
        while (r < s.size()) {
            if (cnt[s[r]]++ == 0) {
                dist++;
            }
            while (dist > k) {
                if(--cnt[s[l++]] == 0) {
                    dist--;
                }
            }
            maxlen = max(maxlen, r - l + 1);
            r++;
        }
        return maxlen;
        
    }
};

File: Sean的leetcode/python/Longest Substring with At Most K Distinct Characters.py
###
# Sliding Window - Memorize index
# Time Complexity: O(n)
# Space Complexity: O(k)
###
class Solution(object):
    def lengthOfLongestSubstringKDistinct(self, s, k):
        """
        :type s: str
        :type k: int
        :rtype: int
        """
        if k == 0 or not s:
            return 0
        maxlen = 0
        l = 0
        d = collections.OrderedDict()
        for r, e in enumerate(s):
            if e in d:
                del d[e]
            d[e] = r
            if len(d) > k:
                _, idx = d.popitem(last=False)
                l = max(l, idx + 1)
            maxlen = max(maxlen, r - l + 1)
        return maxlen
###
# Sliding Window - Memorize counter
# Time Complexity: O(n)
# Space Complexity: O(k)
###
class Solution(object):
    def lengthOfLongestSubstringKDistinct(self, s, k):
        """
        :type s: str
        :type k: int
        :rtype: int
        """
        if k == 0 or not s:
            return 0
        maxlen = 0
        l = 0
        d = collections.Counter()
        for r, e in enumerate(s):
            d[e] += 1
            while len(d) > k:
                d[s[l]] -= 1
                if d[s[l]] == 0:
                    del d[s[l]]
                l += 1
            maxlen = max(maxlen, r - l + 1)
        return maxlen

        
        